nixos/modules/services/web-apps/rimgo.nix

```diff
@@ -0,0 +1,55 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+with lib; let
+  cfg = config.services.rimgo;
+in {
+  options.services.rimgo = {
+    enable = mkEnableOption "rimgo";
+    package = mkPackageOption pkgs "rimgo" {};
+    settings = mkOption {
+      type = with types; attrsOf str;
+      default = {};
+      example = literalExpression ''
+        {
+          FORCE_WEBP = "1";
+        }
+      '';
+      description = ''
+        Settings for rimgo, see [here](https://rimgo.codeberg.page/docs/usage/configuration/) for supported options.
+      '';
+    };
+    port = mkOption {
+      type = types.port;
+      default = 3000;
+      example = 69420;
+      description = "The port to use";
+    };
+    address = mkOption {
+      type = types.str;
+      default = "127.0.0.1";
+      description = "The address to listen on";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    services.rimgo.settings = {
+      PORT = mkDefault (toString cfg.port);
+      ADDRESS = mkDefault cfg.address;
+    };
+    systemd.services.rimgo = {
+      description = "Rimgo";
+      wantedBy = ["default.target"];
+      partOf = ["default.target"];
+      environment = cfg.settings;
+      serviceConfig.ExecStart = "${cfg.package}/bin/rimgo";
```

Comment:
```suggestion
      serviceConfig = {
        ExecStart = "${cfg.package}/bin/rimgo";
        AmbientCapabilities = mkIf (cfg.port < 1024) [ "CAP_NET_BIND_SERVICE" ];
        DynamicUser = true;
        Restart = "on-failure";
        RestartSec = "5s";
        CapabilityBoundingSet =
          if (cfg.port < 1024) then [ "CAP_NET_BIND_SERVICE" ] else [ "" ];
        DeviceAllow = [ "" ];
        LockPersonality = true;
        MemoryDenyWriteExecute = true;
        PrivateDevices = true;
        PrivateUsers = cfg.port >= 1024;
        ProcSubset = "pid";
        ProtectClock = true;
        ProtectControlGroups = true;
        ProtectHome = true;
        ProtectHostname = true;
        ProtectKernelLogs = true;
        ProtectKernelModules = true;
        ProtectKernelTunables = true;
        ProtectProc = "invisible";
        RestrictAddressFamilies = [ "AF_INET" "AF_INET6" ];
        RestrictNamespaces = true;
        RestrictRealtime = true;
        RestrictSUIDSGID = true;
        SystemCallArchitectures = "native";
        SystemCallFilter = [ "@system-service" "~@privileged" ];
        UMask = "0077";
      };
```
This service does not need to run as root with full privileges, so it should run as unprivileged user in a tight sandbox.

Reply:
thanks, that's a lot of stuff i don't understand
i think i'm going to test this first

Reply:
runs wonderfully on rimgo.quantenzitrone.eu :+1:

Reply:
One question: why is `"CAP_NET_BIND_SERVICE"` set with `mkIf` in `AmbientCapabilities`, but with `[""]` as default in `CapabilityBoundingSet`? Wouldn't it be easier to use `mkIf` there too, or is there a specific reason for having the `CapabilityBoundingSet=` line in the resulting service file?

Reply:
`AmbientCapabilities` are additional permissions, and having `CAP_NET_BIND_SERVICE` is required to bind to a low-numbered port. `CapabilityBoundingSet` is a restriction: having the empty string here means that the service cannot obtain any new capabilities, no matter how; not having the option at all does not apply any restriction, thus weakening the sandbox (you can also check this out in `systemd-analyze security`).